### PR TITLE
Ble message receiving

### PR DESCRIPTION
**Pull request by avereha**
Implement LTK: https://github.com/0pen-dash/AndroidAPS/wiki/LTK
 - join and read messages
 - generate the keys
 - implement the rest of the commands:
   - [X] SPS1
   - [x] SPS2
   - [x] SP0GP0, P0
   

Logs on the fake pod
```
INFO[0067] LTK 21f4a9d825ce5e57bad5b4958c6ff95e         
```

and in Logcat:
```
 [OmnipodDashBleManagerImpl.connect():86]: Got LTK: 21f4a9d825ce5e57bad5b4958c6ff95e
```

### CODE REVIEW

File: omnipod-dash/build.gradle [hunk 1]
```diff
@@ -17,4 +17,6 @@ android {
 dependencies {
     implementation project(':core')
     implementation project(':omnipod-common')
-}
\ No newline at end of file
+
+    implementation 'com.google.crypto.tink:tink-android:1.5.0'
```
Comment (bartsopers) [inline omnipod-dash/build.gradle:21]: could you please add the dependency to core_dependencies instead, and add a variable for the version in the root build.gradle?

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/message/CrcMismatchException.kt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.comm.message
+
+import info.nightscout.androidaps.utils.extensions.toHex
+
+class CrcMismatchException(val expected: Long, val got: Long, val payload: ByteArray) :
```
Comment (bartsopers) [inline omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/message/CrcMismatchException.kt:5]: nit: I would rename `got` to `actual`

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/pair/PairResult.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.comm.pair
+
+data class PairResult(val ltk: ByteArray) {
+    init {
+        require(ltk.size == 16)
```
Comment (bartsopers) [inline omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/pair/PairResult.kt:5]: I would add a message here

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/exceptions/MessageIOException.kt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.comm.exceptions
+
+class MessageIOException : Exception {
+    constructor(msg: String) : super(msg)
+    constructor(cause: Throwable) : super(cause)
```
Comment (bartsopers) [inline omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/exceptions/MessageIOException.kt:5]: I would always add a message to an exception, so in this case I would probably do something like `super("Caught Exception during message I/O", cause)`

Comment (avereha) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/exceptions/MessageIOException.kt:5]: Done

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/pair/PairResult.kt (hunk 3 above)
Comment (avereha) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/pair/PairResult.kt:5]: Done

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/message/CrcMismatchException.kt (hunk 2 above)
Comment (avereha) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/message/CrcMismatchException.kt:5]: Done!

File: omnipod-dash/build.gradle (hunk 1 above)
Comment (avereha) [reply omnipod-dash/build.gradle:25]: Done.

File: core/core_dependencies.gradle [hunk 5]
```diff
@@ -46,6 +46,7 @@ dependencies {
 
     //CryptoUtil
     api 'com.madgag.spongycastle:core:1.58.0.0'
+    api "com.google.crypto.tink:tink-android:$tink_version"
```
Comment (AdrianLxM) [inline core/core_dependencies.gradle:49]: 👍

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/pair/LTKExchanger.kt [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+        // TODO: failing to read or validate p0 will lead to undefined state
+        // It could be that:
+        // - the pod answered with p0 and we did not receive/could not process the answer
+        // - the pod answered with some sort of error
+        // But if sps2 conf value is incorrect, then we would probablysee this when receiving the pod podSps2(to test)
+        val p0 = msgIO.receiveMessage()
+        validateP0(p0)
+
+        return PairResult(
+            ltk = ltk,
```
Comment (AdrianLxM) [inline omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/pair/LTKExchanger.kt:78]: nitpick: trailing comma.

**Review by AdrianLxM — APPROVED** [review]
looks good from only looking at the code

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/pair/LTKExchanger.kt (hunk 6 above)
Comment (avereha) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/pair/LTKExchanger.kt]: Done, there were a few others like this in this file.

**Pull request merged**